### PR TITLE
recompile fx.GraphModule lazily

### PR DESCRIPTION
**Pull request by shunting314**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #105257

Context: @eellison 's review comment [here](https://github.com/pytorch/pytorch/pull/103642#discussion_r1244418586) complains about my code calling `torch.fx.GraphModule.recompile` after I changed the graph.  We didn't simply remove the call to `recompile` at that time since that increases the risk that user see or run stale python code. In this PR, I recompile GraphModule lazily without increasing the risk that user see/run stale python code.

When training BertForMaskedLM, the `GraphModule.recompile` is called 707 times and takes 1.8s in total. The whole compilation takes around 60 seconds.

By spot checking, I found the main reason we call recompile so frequently is due to inductor pattern matcher. E.g., if we want to replace src_fn with dst_fn, we need trace both src_fn and dst_fn. After tracing is done, we create a GraphModule. The init method of GraphModule will call recompile.

By doing recompile lazily, we reduce the number of calls for `GraphModule._real_recompile` (in this PR, `recompile` just mark the class as needing recompilation and is very light weight. `_real_recompile` does the real recompilation) to 37 times and reduces its total execution time to 0.045s.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng @avikchaudhuri @gmagogsfm @zhxchen17 @tugsbayasgalan @angelayi @suo @ydwu4 @Xia-Weiwen @anijain2305 @ipiszy

### CODE REVIEW

**Comment by ezyang** [comment]
I'll review after tests are passing

**Review by jansel — CHANGES_REQUESTED** [review]
Tests seem pretty broken.

Rather than hooking into `__call__` (which won't work if one calls `forward` directly), I'd suggest:

```
def _lazy_forward(self, *args, **kwargs):
   self._real_recompile()
   assert not self._needs_recompile()
   return self.forward(*args, **kwargs)

forward = _lazy_forward

@classmethod
def _needs_recompile(cls):
   return cls.forward is cls._lazy_forward

@classmethod
def recompile(cls):
   cls.forward = cls._lazy_forward
```

**Comment by jansel** [comment]
Actually, you might not want the `@classmethod` bits.  It looks like this is modifying `self.foward` not `cls.forward`.

**Comment by shunting314** [comment]
> Actually, you might not want the @classmethod bits. It looks like this is modifying self.foward not cls.forward.

The `__new__` method (https://github.com/pytorch/pytorch/blob/main/torch/fx/graph_module.py#L300) of GraphModule is implemented in such a way that every new instance has its own GraphModuleImpl subclass define. So changing `forward` on the GraphModuleImpl class will only impact that specific instance -- the behavior is like we change self.forward.

I think this 'a separate subclass per GraphModule instance' trick is not 100% necessary. An alternative is as you said, we change `forward` method on each GraphModule instance rather than on the subclass. But since our code already has such a design, maybe we should make my change consistent with the current design for now. If we want, I can do a refactoring as a followup. Not sure if there is other reason that we pick such a design in the first place though.

File: test/fx/test_lazy_recompile.py [hunk 1]
```diff
@@ -0,0 +1,57 @@
+# Owner(s): ["oncall: fx"]
+
+from torch.testing._internal.common_utils import TestCase, run_tests
+from torch import fx
+import torch
+import torch._export
+
+class TestLazyRecompile(TestCase):
```
Comment (jansel) [inline test/fx/test_lazy_recompile.py:8]: 1) Add a test where you call, mutate the graph, and call again, so multiple recompiles are needed.

2) Add a test that asserts `needs_recompile()` is False after the compile happens (and true before it)

**Review by jansel — CHANGES_REQUESTED** [review]

File: torch/utils/_sympy/value_ranges.py [hunk 2]
```diff
@@ -540,7 +540,7 @@ def where(a, b, c):
             return ValueRanges(sympy.Min(b.lower, c.lower), sympy.Max(b.upper, c.upper))
 
     def __getattr__(self, name):
-        log.warning("unhandled ValueRange op %s", name)
+        log.debug("unhandled ValueRange op %s", name)
```
Comment (ezyang) [inline torch/utils/_sympy/value_ranges.py:543]: How come?

File: torch/_functorch/compilers.py [hunk 3]
```diff
@@ -183,6 +183,10 @@ def debug_nop(fx_g: fx.GraphModule, _) -> Callable:
 @make_boxed_compiler
 def simple_ts_compile(fx_g, _):
     strip_overloads(fx_g)
+
+    # realize the lazy recompilication to make jit.script happy.
+    fx_g.real_recompile()
```
Comment (ezyang) [inline torch/_functorch/compilers.py:188]: This looks suspicious. Are you breaking BC here?

File: torch/utils/_sympy/value_ranges.py (hunk 2 above)
Comment (shunting314) [reply torch/utils/_sympy/value_ranges.py:543]: So running the benchmark script for BertForMaskedLM will execute this line a lot complaining about random seed related attributes. Those looks not actionable so I suppress the warning here to avoid flooding the output.  Maybe we can change this log to warning again once the underlying problem is fixed.

File: torch/_functorch/compilers.py (hunk 3 above)
Comment (shunting314) [reply torch/_functorch/compilers.py:188]: I change the behavior that `GraphModule.recompile` will not do the recompilation right away. The recompilation will only be triggered later when we call `__call__`/`forward` or access properties like `code`.

This indeed cause some trouble for torchscript here. Without the ***real*** recompilation, the torchscript will see our ***lazy*** version of the `forward` method and fail. So I force feeding the ***recompiled*** version of `forward` method to torchscript. I think there can be a more transparent way to do this. But I'm not quite familiar with torchscript and not sure where is a good place that I can add code like:

```
if isinstance(mod, fx.GraphModule):
  mod.real_recompile()
```

File: torch/fx/graph_module.py [hunk 4]
```diff
@@ -642,12 +642,50 @@ def code(self) -> str:
...
 
+    @property
+    def in_spec(self):
+        self.real_recompile()
+        return getattr(self, "_in_spec", None)
+
+    @property
+    def out_spec(self):
+        self.real_recompile()
+        return getattr(self, "_out_spec", None)
```
Comment (ezyang) [inline torch/fx/graph_module.py:659]: This feels extra awkward. It seems like you have a BC-breaking change (previous access to `_out_spec` is now `out_spec`) PLUS you won't see anything is wrong until it breaks! If you *must* break it, best to rename the old underscored fields to something else!

File: torch/fx/graph_module.py [hunk 5]
```diff
@@ -642,12 +642,50 @@ def code(self) -> str:
...
+    def _lazy_forward(self, *args, **kwargs):
+        self._real_recompile()
+        assert not self._needs_recompile()
+        return self.forward(*args, **kwargs)
+
+    forward = _lazy_forward
+
+    def real_recompile(self):
+        """
+        A torch script safe wrapper around _real_recompile.
```
Comment (ezyang) [inline torch/fx/graph_module.py:679]: It's incredibly strange to me that recompile is being called from TorchScriptable code!

**Review by eellison — COMMENTED** [review]
If this gets hard to land you might have better luck with adding a lazy-recompile method which we call inside torch/_inductor / PT2 stack, instead of trying to always make recompile() always lazy. As you said, it's mostly within the pattern matcher that the invocations come from.

File: torch/fx/graph_module.py (hunk 5 above)
Comment (shunting314) [reply torch/fx/graph_module.py:679]: Here is how it happens from looking at some failure stack trace. When torchscript gets a GraphModule, it will try to inspect the GraphModule's attributes. When it looks at the 'code' property, it will try to 'script' the method implementation. Due to the lazy recompilation, we may need to recompile the graph module when 'code' property is being accessed. That's how what you described happens.

I think a not too bad fix is to always force a ***real*** recompilation when TorchScript backend gets a graph module.

File: torch/fx/graph_module.py (hunk 4 above)
Comment (shunting314) [reply torch/fx/graph_module.py:659]: I totally agree :) The reason that I do this BC breaking change is because I think
1. '_in_spec' / '_out_spec' is some private fields that normal user is not supposed to use
2. there are only very small number of usages of them 

I can keep BC compatible by changing `_in_spec` into a property and let it access an attributed called something like `__in_spec` (double underscore) . But an alternative is we just create an property called `in_spec` and change the places previously accessing `_in_spec` to access `in_spec` instead. My assumption is `_in_spec`/`_out_spec` does not belong to public APIs, so doing some BC breaking change should be fine. Let me know if I'm wrong. I'll be glad to change my implementation

Comment (shunting314) [reply torch/fx/graph_module.py:659]: I think `renaming` is a good idea. I can do something like:
1. create a field `_graph_in_spec`/`_graph_out_spec` that serves the same purpose as what `_in_spec`/`_out_spec` is doing
2. create property `graph_in_spec`/`graph_out_spec` to access _graph_in_spec and _graph_out_spec and do recompilation as necessary
3. change `_in_spec`/`_out_spec` to alias to `graph_in_spec`/`graph_out_spec` and deprecate their usage. (Q: what API should I use to deprecate a property?)

File: torch/fx/graph_module.py (hunk 5 above)
Comment (shunting314) [reply torch/fx/graph_module.py:679]: Ideally yes. But this hits some limitation of torchscript - some of the lazy recompile related methods can not be handled by torchscript. Since we are deprecating torchscript, probably we should workaround rather than spending efforts to improve torchscript?

**Comment by shunting314** [comment]
I roughly went thru the remaining failed tests. Here are the main failures:
- inline in skipfiles: GraphModule._lazy_forward
https://github.com/pytorch/pytorch/actions/runs/5582213089/jobs/10201953075?pr=105257 
- test_fx_experimental:
https://github.com/pytorch/pytorch/actions/runs/5582213089/jobs/10201492606?pr=105257 
- functorch test_ops:
https://github.com/pytorch/pytorch/actions/runs/5582213089/jobs/10201495008?pr=105257 
- GraphModule has no _code attribute:
https://github.com/pytorch/pytorch/actions/runs/5582213089/jobs/10201932881?pr=105257 

I'll try to fix them. But if things go out of control, I think we can make lazy recompile optional and disable by default (as Elias also mentioned earlier).

File: torch/fx/graph_module.py (hunk 5 above)
Comment (shunting314) [reply torch/fx/graph_module.py:679]: @ezyang  so we encounter similar issues when passing a lazy recompiled GraphModule to dynamo. Similar to the torchscript case, there is an option to enhance the graph capturing layer (dynamo or torchscript) to handle the GraphModule lazy recompilation. But that may take quite some effort and I'm not sure if it's necessary. it would be handy to just force the recompile at the beginning of these graph capturing layers. Wdyt?

**Review by zdevito — COMMENTED** [review]
I'm not sure changing the (currently pretty simple) behavior of GraphModule is the best approach here. If the problem is that the torch compile pathway is it wants to make many changes to GraphModule and then update when it is done, maybe the logic that handles it should go in that pathway rather than GraphModule. For instance, can it use a subclass of GraphModule that has the lazy behavior?

**Comment by shunting314** [comment]
@zdevito  thanks for the comments.

First let me provide more context to this problem. From on hand, we want to change a graph multiple times and recompile once as you mentioned. From the other hand, it happens quite frequently in inductor that we create a GraphModule just for its graph and nodes. We never run or check the source code. So recompiling is a waste.

IIUC, you comments want 2 things
1. make lazy recompile optional (only affect the pathway that wants it)
2. don't change GraphModule API but use a subclass

For the first item, we've also being talking about this earlier. And the current plan is we do that if enabling lazy recompile is too hard to make it work correctly.

For the 2nd item, I feel it may be harder if we don't do it in the GraphModule class. The reason is, there are quite a lot of places that will create a GraphModule. Here are a few places that inductor may run into:
- in make_fx called directly in inductor or called by aot_autograd: https://github.com/pytorch/pytorch/blob/main/torch/_functorch/aot_autograd.py#L1399
- in the partitioner related API: https://github.com/pytorch/pytorch/blob/main/torch/_functorch/partitioners.py#L219-L220

I'm not sure if there are more.

We would need to add code to pick lazy or original version of GraphModule in all these places.

**Comment by shunting314** [comment]
I've fixed most of the test failures. The major remaining one is failing test_function_back_compat in test_fx.py. The reason is, previously `recompile` returns a `PythonCode` and now since we do lazy recompilation, we don't have `PythonCode` to return and we return None instead.

I'm not sure if this BC issue is a major one. If we need 100% be backward compatible, I can do this:
- keep the original `recompile` untouched
- create a new `lazy_recompile` API that do lazy recompilation.
- in a follow up PR, change all calls of `recompile` to `lazy_recompile` in inductor

Would like to know people's thoughts on this. @jansel @ezyang @eellison @zdevito

**Comment by jansel** [comment]
I think we need to be backwards compatible.  Let's create a `LazyGraphModule` subclass that we can use in dynamo/inductor.
